google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -0,0 +1,364 @@
+    this.streamConnection =
+        new StreamConnection(
+            builder.client,
+            new RequestCallback() {
+              @Override
+              public void run(AppendRowsResponse response) {
+                requestCallback(response);
+              }
+            },
+            new DoneCallback() {
+              @Override
+              public void run(Throwable finalStatus) {
+                doneCallback(finalStatus);
+              }
+            });
+    this.appendThread =
+        new Thread(
+            new Runnable() {
+              @Override
+              public void run() {
+                appendLoop();
+              }
+            });
+    this.appendThread.start();
+  }
+
+  /**
+   * Schedules the writing of a message.
+   *
+   * <p>Example of writing a message.
+   *
+   * <pre>{@code
+   * AppendRowsRequest message;
+   * ApiFuture<AppendRowsResponse> messageIdFuture = writer.append(message);
+   * ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<AppendRowsResponse>() {
+   *   public void onSuccess(AppendRowsResponse response) {
+   *     if (response.hasOffset()) {
+   *       System.out.println("written with offset: " + response.getOffset());
+   *     } else {
+   *       System.out.println("received an in stream error: " + response.error().toString());
+   *     }
+   *   }
+   *
+   *   public void onFailure(Throwable t) {
+   *     System.out.println("failed to write: " + t);
+   *   }
+   * }, MoreExecutors.directExecutor());
+   * }</pre>
+   *
+   * @param message the message in serialized format to write to BigQuery.
+   * @return the message ID wrapped in a future.
+   */
+  public ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    AppendRequestAndResponse requestWrapper = new AppendRequestAndResponse(message);
+    this.lock.lock();
+    try {
+      if (userClosed) {
+        requestWrapper.appendResult.setException(
+            new StatusRuntimeException(
+                Status.fromCode(Status.Code.FAILED_PRECONDITION)
+                    .withDescription("Stream is already closed")));
+        return requestWrapper.appendResult;
+      }
+      if (connectionFinalStatus != null) {
+        requestWrapper.appendResult.setException(
+            new StatusRuntimeException(
+                Status.fromCode(Status.Code.FAILED_PRECONDITION)
+                    .withDescription(
+                        "Stream is closed due to " + connectionFinalStatus.toString())));
+        return requestWrapper.appendResult;
+      }
+      waitingRequestQueue.addLast(requestWrapper);
+      hasMessageInWaitingQueue.signal();
+      return requestWrapper.appendResult;
+    } finally {
+      this.lock.unlock();
+    }
+  }
+
+  /** Close the stream writer. Shut down all resources. */
+  @Override
+  public void close() {
+    log.info("User closing stream: " + streamName);
+    this.lock.lock();
+    try {
+      this.userClosed = true;
```

Comment:
adding a local var to remember the previous value, so if user closed twice, the second time will be no_op (don't call the appendThread.join() multiple times).